(WIP) Fix retain grad behavior on in-place

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #79767
* #79902

Registering a hook to retain grad involves dealing with two layers of hooks. First there is a hook list that is stored on the autograd meta, and then there is a single prehook registered on the grad_fn that invokes all the hooks in this hook list. 

Previously, when an in-place happens, the prehook registered to the old grad_fn still references the hook list on the tensors autograd meta. This PR makes it so that on-rebase we remove the prehook registered to the old grad_fn and creates a new prehook that on the new grad_fn referencing the same hook list on autograd meta.

Not sure if this is the right fix though since now hooks registered from python and the retain grad hook behave differently on in-place. Perhaps a similar update needs to be made for the register_hooks Python api?
```
a = torch.tensor([1.], requires_grad=True).clone()

grad1 = [None]
grad2 = [None]

def hook1(grad):
    grad1[0] = grad

def hook2(grad):
    grad2[0] = grad

a.register_hook(hook1)
a.retain_grad()

a.mul_(2)

b = a.clone()
b.register_hook(hook2)

out = (b + 1).sum()
out.backward()

print(grad1)  # hook before 2.0
print(grad2)  # hook after  1.0
print(a.grad) # retain_grad 1.0
```
Other TODO:
- Add tests
- Consider the in-place over view case
- Remove retain grad after backward with inputs (will be done in follow up)

soulitzer committed:
(WIP) Fix retain grad behavior on in-place

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/variable.h:
```diff
@@ -225,6 +225,10 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
   // Only meaningful on non-leaf variables (must be false otherwise)
   bool retains_grad_;
 
+  // When CppFunctionPreHook is registered as a prehook on the grad_fn,
```
Not anymore

soulitzer commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -181,19 +181,29 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
       new CppFunctionPreHook(list, self.output_nr()));
-  clear_hooks(self);
+  if (!use_existing_hook_list) {
+    clear_hooks(self);
```
Why do we also need to register the hooks to this list?

soulitzer commented on torch/csrc/autograd/function.h:
```diff
@@ -426,8 +426,21 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
   }
 
-  void add_pre_hook(std::unique_ptr<FunctionPreHook>&& pre_hook) {
+  uintptr_t add_pre_hook(std::unique_ptr<FunctionPreHook>&& pre_hook) {
```
These are just copies of the add_post_hook and del_post_hook above

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -181,19 +181,29 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
   }
 
   set_gradient_edge(self, std::move(gradient_edge));
+  if (self.retains_grad()) {
```
There can be early return in this functions so this is not enough.

soulitzer replied:
Yup, will indeed need to handle the in-place over view case

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -181,19 +181,29 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
 
   set_gradient_edge(self, std::move(gradient_edge));
+  if (self.retains_grad()) {
+    auto old_fn = self.grad_fn()->next_edges()[0].function;
```
Why not do that before the grad_fn() is updated? That way you don't have to traverse the edge.

soulitzer replied:
Good catch! This would probably simplify the in-place over view case as well

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -181,19 +181,29 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
   set_gradient_edge(self, std::move(gradient_edge));
+  if (self.retains_grad()) {
+    auto old_fn = self.grad_fn()->next_edges()[0].function;
+    old_fn->del_pre_hook(materialize_autograd_meta(self)->grad_fn_prehook_key);
```
This actually remove all the hooks that were registered via this Tensor, not just the retain_grad one right?

soulitzer replied:
Yes all hooks that were registered with the cpp register_hooks API

soulitzer commented:
@pytorchbot rebase

pytorchmergebot committed:
Update on "(WIP) Fix retain grad behavior on in-place"


Registering a hook to retain grad involves dealing with two layers of hooks. First there is a hook list that is stored on the autograd meta, and then there is a single prehook registered on the grad_fn that invokes all the hooks in this hook list. 

Previously, when an in-place happens, the prehook registered to the old grad_fn still references the hook list on the tensors autograd meta. This PR makes it so that on-rebase we remove the prehook registered to the old grad_fn and creates a new prehook that on the new grad_fn referencing the same hook list on autograd meta.

Not sure if this is the right fix though since now hooks registered from python and the retain grad hook behave differently on in-place. Perhaps a similar update needs to be made for the register_hooks Python api?
```
a = torch.tensor([1.], requires_grad=True).clone()

grad1 = [None]
grad2 = [None]

def hook1(grad):
    grad1[0] = grad

def hook2(grad):
    grad2[0] = grad

a.register_hook(hook1)
a.retain_grad()

a.mul_(2)

b = a.clone()
b.register_hook(hook2)

out = (b + 1).sum()
out.backward()

print(grad1)  # hook before 2.0
print(grad2)  # hook after  1.0
print(a.grad) # retain_grad 1.0
```
Other TODO:
- Add tests
- Consider the in-place over view case
- Remove retain grad after backward with inputs (will be done in follow up)

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Fix retain grad behavior on in-place"


Registering a hook to retain grad involves dealing with two layers of hooks. First there is a hook list that is stored on the autograd meta, and then there is a single prehook registered on the grad_fn that invokes all the hooks in this hook list. 

Previously, when an in-place happens, the prehook registered to the old grad_fn still references the hook list on the tensors autograd meta. This PR makes it so that on-rebase we remove the prehook registered to the old grad_fn and creates a new prehook that on the new grad_fn referencing the same hook list on autograd meta.

Not sure if this is the right fix though since now hooks registered from python and the retain grad hook behave differently on in-place. Perhaps a similar update needs to be made for the register_hooks Python api?
```
a = torch.tensor([1.], requires_grad=True).clone()

grad1 = [None]
grad2 = [None]

def hook1(grad):
    grad1[0] = grad

def hook2(grad):
    grad2[0] = grad

a.register_hook(hook1)
a.retain_grad()

a.mul_(2)

b = a.clone()
b.register_hook(hook2)

out = (b + 1).sum()
out.backward()

print(grad1)  # hook before 2.0
print(grad2)  # hook after  1.0
print(a.grad) # retain_grad 1.0
```
Other TODO:
- Add tests
- Consider the in-place over view case
- Remove retain grad after backward with inputs (will be done in follow up)

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Fix retain grad behavior on in-place"


Registering a hook to retain grad involves dealing with two layers of hooks. First there is a hook list that is stored on the autograd meta, and then there is a single prehook registered on the grad_fn that invokes all the hooks in this hook list. 

Previously, when an in-place happens, the prehook registered to the old grad_fn still references the hook list on the tensors autograd meta. This PR makes it so that on-rebase we remove the prehook registered to the old grad_fn and creates a new prehook that on the new grad_fn referencing the same hook list on autograd meta.

Not sure if this is the right fix though since now hooks registered from python and the retain grad hook behave differently on in-place. Perhaps a similar update needs to be made for the register_hooks Python api?
```
a = torch.tensor([1.], requires_grad=True).clone()

grad1 = [None]
grad2 = [None]

def hook1(grad):
    grad1[0] = grad

def hook2(grad):
    grad2[0] = grad

a.register_hook(hook1)
a.retain_grad()

a.mul_(2)

b = a.clone()
b.register_hook(hook2)

out = (b + 1).sum()
out.backward()

print(grad1)  # hook before 2.0
print(grad2)  # hook after  1.0
print(a.grad) # retain_grad 1.0
```
Other TODO:
- Add tests
- Consider the in-place over view case
- Remove retain grad after backward with inputs (will be done in follow up)

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Fix retain grad behavior on in-place"


Registering a hook to retain grad involves dealing with two layers of hooks. First there is a hook list that is stored on the autograd meta, and then there is a single prehook registered on the grad_fn that invokes all the hooks in this hook list. 

Previously, when an in-place happens, the prehook registered to the old grad_fn still references the hook list on the tensors autograd meta. This PR makes it so that on-rebase we remove the prehook registered to the old grad_fn and creates a new prehook that on the new grad_fn referencing the same hook list on autograd meta.

Not sure if this is the right fix though since now hooks registered from python and the retain grad hook behave differently on in-place. Perhaps a similar update needs to be made for the register_hooks Python api?
```
a = torch.tensor([1.], requires_grad=True).clone()

grad1 = [None]
grad2 = [None]

def hook1(grad):
    grad1[0] = grad

def hook2(grad):
    grad2[0] = grad

a.register_hook(hook1)
a.retain_grad()

a.mul_(2)

b = a.clone()
b.register_hook(hook2)

out = (b + 1).sum()
out.backward()

print(grad1)  # hook before 2.0
print(grad2)  # hook after  1.0
print(a.grad) # retain_grad 1.0
```
Other TODO:
- Add tests
- Consider the in-place over view case
- Remove retain grad after backward with inputs (will be done in follow up)

[ghstack-poisoned]

soulitzer commented:
Closing in favor of https://github.com/pytorch/pytorch/pull/79996
